src/lfortran/semantics/ast_common_visitor.h

```diff
@@ -8088,6 +8088,69 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
+            bool check_m_m = true; 
+            if(ASR::is_a<ASR::ArrayItem_t>(*ASRUtils::EXPR(tmp_copy))){
+                tmp2 = ASR::down_cast<ASR::StructInstanceMember_t>(ASR::down_cast<ASR::ArrayItem_t>(ASRUtils::EXPR(tmp_copy))->m_v);
+                check_m_m = false;
+            } else if (ASR::is_a<ASR::StructInstanceMember_t>(*ASRUtils::EXPR(tmp_copy))) {
+                tmp2 = ASR::down_cast<ASR::StructInstanceMember_t>(ASRUtils::EXPR(tmp_copy));
+            }
+
+            if(check_m_m){
+                ASR::ExternalSymbol_t* tmp2_m_m_ext = ASR::down_cast<ASR::ExternalSymbol_t>(tmp2->m_m);
+                if(ASR::is_a<ASR::Variable_t>(*(tmp2_m_m_ext->m_external)) && 
+                    ASRUtils::type_get_past_allocatable(ASRUtils::symbol_type(tmp2_m_m_ext->m_external))->type == ASR::ttypeType::Array){
+                    if(array_found){
+                        throw SemanticError("Two or more part references with non-zero rank must not be specified.", loc);
+                    }
+                    array_found = true;
+                    array_type = ASRUtils::duplicate_type(al,ASRUtils::symbol_type(tmp2->m_m));                        
+                }
+            }
+            if(tmp2->m_v->type == ASR::exprType::Var){
+                ASR::ttype_t* var_type = ASRUtils::expr_type(tmp2->m_v);
+                if(ASRUtils::type_get_past_allocatable(var_type)->type == ASR::ttypeType::Array){
+                    if(array_found){
+                        throw SemanticError("Two or more part references with non-zero rank must not be specified.", loc);
+                    }
+                    array_found = true;
+                    array_type = ASRUtils::duplicate_type(al,var_type);
+                }
+            }
+            tmp_copy = (ASR::asr_t*)(tmp2->m_v);
+        }
+        if(array_type){
+            if(ASR::is_a<ASR::StructInstanceMember_t>(*ASRUtils::EXPR(tmp))){
+                ASR::StructInstanceMember_t* tmp2 = ASR::down_cast<ASR::StructInstanceMember_t>(ASRUtils::EXPR(tmp)); 
+                if(ASR::is_a<ASR::Array_t>(*array_type)){
+                    ((ASR::Array_t*)array_type)->m_type = ASRUtils::type_get_past_array(tmp2->m_type);
+                    tmp2->m_type = array_type;
+                }
+                if(ASR::is_a<ASR::Allocatable_t>(*array_type)){
+                    ASR::down_cast<ASR::Array_t>(((ASR::Allocatable_t*)array_type)->m_type)->m_type = ASRUtils::type_get_past_array(ASRUtils::type_get_past_allocatable(tmp2->m_type));
+                    tmp2->m_type = array_type;
+                }
+                
+            } else if (ASR::is_a<ASR::ArrayItem_t>(*ASRUtils::EXPR(tmp))) {
+                ASR::ArrayItem_t* tmp2 = ASR::down_cast<ASR::ArrayItem_t>(ASRUtils::EXPR(tmp));
+                if(ASR::is_a<ASR::Array_t>(*array_type)){
+                    ((ASR::Array_t*)array_type)->m_type = ASRUtils::type_get_past_array(tmp2->m_type);
```

Comment:
Use `down_cast` for every downcast:
```suggestion
                    ASR::down_cast<ASR::Array_t>(array_type)->m_type = ASRUtils::type_get_past_array(tmp2->m_type);
```
And the same at all the other places above. This is checked in Debug mode.